src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/LanguageServiceHost.cs

```diff
@@ -0,0 +1,145 @@
+            Requires.NotNull(activeConfiguredProjectSubscriptionService, nameof(activeConfiguredProjectSubscriptionService));
+
+            _commonServices = commonServices;
+            _contextProvider = contextProvider;
+            _tasksService = tasksService;
+            _activeConfiguredProjectSubscriptionService = activeConfiguredProjectSubscriptionService;
+
+            Handlers = new OrderPrecedenceImportCollection<ILanguageServiceRuleHandler>(projectCapabilityCheckProvider: commonServices.Project);
+        }
+
+        [ImportMany]
+        public OrderPrecedenceImportCollection<ILanguageServiceRuleHandler> Handlers
+        {
+            get;
+        }
+
+        [ProjectAutoLoad(ProjectLoadCheckpoint.ProjectFactoryCompleted)]
+        [AppliesTo(ProjectCapability.CSharpOrVisualBasicLanguageService2)]
+        private Task OnProjectFactoryCompletedAsync()
+        {
+            using (_tasksService.LoadedProject())
+            {
+                var watchedEvaluationRules = GetWatchedRules(RuleHandlerType.Evaluation);
+                var watchedDesignTimeBuildRules = GetWatchedRules(RuleHandlerType.DesignTimeBuild);
+
+                _designTimeBuildSubscriptionLink = _activeConfiguredProjectSubscriptionService.JointRuleSource.SourceBlock.LinkTo(
+                  new ActionBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>>(e => OnProjectChanged(e, RuleHandlerType.DesignTimeBuild)),
+                  ruleNames: watchedDesignTimeBuildRules.Union(watchedEvaluationRules), suppressVersionOnlyUpdates: true);
+
+                _evaluationSubscriptionLink = _activeConfiguredProjectSubscriptionService.ProjectRuleSource.SourceBlock.LinkTo(
+                    new ActionBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>>(e => OnProjectChanged(e, RuleHandlerType.Evaluation)),
+                    ruleNames: watchedEvaluationRules, suppressVersionOnlyUpdates: true);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        protected async override Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            // Don't initialize if we're unloading
+            _tasksService.UnloadCancellationToken.ThrowIfCancellationRequested();
+
+            _projectContext = await _contextProvider.Value.CreateProjectContextAsync()
+                                                          .ConfigureAwait(false);
+
+            foreach (var handler in Handlers)
+            {
+                handler.Value.SetContext(_projectContext);
+            }
+        }
+
+        private async Task OnProjectChanged(IProjectVersionedValue<IProjectSubscriptionUpdate> e, RuleHandlerType handlerType)
+        {
+            if (IsDisposing || IsDisposed)
+                return;
+
+            await InitializeAsync().ConfigureAwait(false);
+
+            // TODO: https://github.com/dotnet/roslyn-project-system/issues/353
+            await _commonServices.ThreadingService.SwitchToUIThread();
```

Comment:
We don't receive events in different orders - they are always guaranteed to version increasing. If listening to a single dataflow port they are also guaranteed to never overlap.

We *might* have two events coming in from evaluation and a design-time build, will play around with this.